test/library/packages/Sort/RadixSort/RadixMSB.chpl

```diff
@@ -0,0 +1,93 @@
+ 
+ module RadixMSB {
+   use BitOps;
+   use Sort;
+   use Random;
+   config param bucketBits=8; 
+
+   proc radixSortMSB(array:[?dom] int) {
+
+     var auxArray:[1..array.size] int;
+     var place = findPlace(array);
+     __radixSortMSB(array,place,dom.low,dom.high,auxArray);
+
+   }
+  
+   private proc __radixSortMSB(array:[] int, place: int, startIndex: int, endIndex: int, auxArray:[] int) {
+
+     // comparison sorting for <=100 elements
+     if(array.size<=100) {
+       insertionSort(array);
+       return;
+     }
+     
+     param numBuckets = (1 << bucketBits);
+     var counts:[0..numBuckets] int;
+     var curOffsets: [0..numBuckets] int; 
+   
+     //base case, exit condition
+     if (endIndex<=startIndex || place<0) {
+       return;
+     }
+   
+     for i in startIndex..endIndex {
+       // find the postion to store a particular number in the counts array
+       var currentDigit = findPosition(place, array[i], numBuckets);
+     
+       //increment count at that position
+       counts[currentDigit] += 1; //increment count at the position of the m.s.d.
+     }
+    
+     //update curOffsets
+     curOffsets[0] = 0;
+     for i in 1..numBuckets {
+       curOffsets[i] = curOffsets[i-1] + counts[i-1];
+     }
+
+
+     //fill in the auxarray at correct positions 
+     for i in startIndex..endIndex {
+       var countIndex = findPosition(place, array[i], numBuckets);
+       var auxIndex = curOffsets[countIndex];     
+       auxArray[auxIndex+1] = array[i];   
+       curOffsets[countIndex] += 1;                 
+     }
+   
+     //swap slices of array and aux array   
+     array[startIndex..endIndex] <=> auxArray[1..(endIndex-startIndex)];
+ 
+
+     //recursive call for subparts of array
+     for m in 0..numBuckets-1 {        
+       if m==0 {
+         __radixSortMSB(array, place-bucketBits, startIndex, startIndex + curOffsets[m] -1, auxArray);
```

Comment:
Reason for extra recursive call at m==0: 
For the purpose of this particular example, consider the **base is 10**:

Input array: 1, 5, 16, 3, 21, 5, 3, 1

**Previous method** - going backwards using counts array: 

Indices:  0 1 2 3 4 5 6 7 8 9
Counts:  6 1 1 0 0 0 0 0 0 0
Updated
Counts:  6 7 8 8 8 8 8 8 8 8
After the elements are put into their respective positions according to 1st MSB: 

Updated
Counts:  0 6 7 8 8 8 8 8 8 8 

The recursive calls here are done using the updated counts.
In the first recursive call of the loop, 
 new start_index = start_index + counts[m] = 1 + 0  = 1
 new end_index = end_index + counts[m+1] = 1 + 0  = 1

Next recursive call,
new start_index = 7
 new end_index = 7

**Current method** - going forward using curOffsets:

Indices:       0 1 2 3 4 5 6 7 8 9
Counts:       6 1 1 0 0 0 0 0 0 0
curOffsets:  0 6 7 8 8 8 8 8 8 8 
After the elements are put into their respective positions according to 1st MSB: 

Updated
curOffsets:   6 7 8 8 8 8 8 8 8 8
The recursive calls here are done using the updated curOffsets.

In the first recursive call of the loop, 
 new start_index = start_index + counts[m] = 1 + 6  = 7
 new end_index = end_index + counts[m+1] = 1 + 0  = 7

One recursive call, indicating the first bucket to be sorted, that is with start_index as 1 and end_index as 6, as indicated in the previous method example above, is missed here, so to cater for this, 2 recursive calls are made at m==0.

Reply:
I can think of two ways to maybe make this look nicer:
 1. Put the extra call before the loop
 2. Make curOffsets array start at -1 of do for m in -1..numBuckets-1